internal/buffer_pool.go

```diff
@@ -45,19 +50,107 @@ func maxMemToUse(buffersNow uint64) uint64 {
 		panic(err)
 	}
 
-	log.Debugf("amount of available memory: %v", m.Available/1024/1024)
+    availableMem := getCgroupAvailableMem()
+	log.Debugf("amount of available memory from cgroup is: %v", availableMem/1024/1024)
+
+    if(availableMem < 0 || availableMem > m.Available) {
+        availableMem = m.Available
+    }
+
+	log.Debugf("amount of available memory: %v", availableMem/1024/1024)
 
 	var ms runtime.MemStats
 	runtime.ReadMemStats(&ms)
 
 	log.Debugf("amount of allocated memory: %v %v", ms.Sys/1024/1024, ms.Alloc/1024/1024)
 
-	max := uint64(m.Available+ms.Sys) / 2
+	max := uint64(availableMem+ms.Sys) / 2
 	maxbuffers := MaxUInt64(max/BUF_SIZE, 1)
 	log.Debugf("using up to %v %vMB buffers, now is %v", maxbuffers, BUF_SIZE/1024/1024, buffersNow)
 	return maxbuffers
 }
 
+const CGROUP_PATH = "/proc/self/cgroup"
+const CGROUP_FOLDER_PREFIX = "/sys/fs/cgroup/memory" 
+const MEM_LIMIT_FILE_SUFFIX = "/memory.limit_in_bytes"
+const MEM_USAGE_FILE_SUFFIX = "/memory.max_usage_in_bytes"
+
+func getCgroupAvailableMem() (retVal uint64) {
+    //get the memory cgroup for self and send limit - usage for the cgroup
+
+    //Return this value in error case
+    retVal = math.MaxUint64
+ 
+    data,err := ioutil.ReadFile(CGROUP_PATH)
+    if err != nil {
+        log.Debugf("Unable to read file %s error: %s", CGROUP_PATH, err)
+        return
+    }
+
+    path := getMemoryCgroupPath(string(data)) 
+    if path == "" {
+        log.Debugf("Unable to get memory cgroup path") 
+        return
+    }
+    log.Debugf("the memory cgroup path for the current process is %v", path)  
+
+    mem_limit, err := readFileAndGetValue(filepath.Join(CGROUP_FOLDER_PREFIX, path,  MEM_LIMIT_FILE_SUFFIX))
+    if err != nil {
+        log.Debugf("Unable to get memory limit from cgroup error: %v", err)
+        return
+    }
+
+    mem_usage, err := readFileAndGetValue(filepath.Join(CGROUP_FOLDER_PREFIX, path, MEM_USAGE_FILE_SUFFIX))
+    if err != nil {
+        log.Debugf("Unable to get memory usage from cgroup error: %v", err)
+        return
+    }
+
+    return (mem_limit - mem_usage)
```

Comment:
mem_usage is  `memory.max_usage_in_bytes	 # show max memory usage recorded` according to https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt , which sounds like it will show the peak usage even if the current usage is less? seems to me that you want to get the memory limit from cgroup and check the current usage the old fashion way

Reply:
yes usage_in_bytes sounds more correct than max_usage_in_bytes. I don't mind that we pull in an external dependency for this